### PR TITLE
Enabling a new DevKit build job and support use within pipeline configuration

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
Create a ci-jenkins-pipelines "Build DevKit" job to build gcc Linux based DevKits https://openjdk.org/groups/build/doc/building.html
These jobs perform:
1. Clone the required openjdk source containing the "make/devkit" make files
2. patch the Tools.gmk to enable building Centos7 based DevKit's required by Eclipse Adoptium to build Temurin.
3. Tools.gmk patch also add GPG download verification
4. binutils-2.39.patch makefile patch required to enable building on Centos7
5. Builds the DevKit...
6. GPG Signs the compressed archive
7. Performs a "Publish DevKit" "DryRun" - Which can then be re-run to actually publish if required to https://github.com/adoptium/devkit-binaries/releases

DevKit jobs using this PR actively building DevKits for jdk21+ gcc 11.3.0 x64/aarch64/ppc64le linux : https://ci.adoptium.net/job/build-scripts/job/utils/job/devkit/job/gcc-11.3.0/

openjdk_build_pipeline support for DevKit in jdk21+:
1. Use user openjdk build directory, so that debug symbol folders are mapped reproducibly
2. Ensure build/devkit folder is cleaned as part of post-build workspace build output clean

### CODE REVIEW

**Comment by sxa** [comment]
A few thoughts:
- Is there a reason why the job name is e.g. `x64Linux` and not `linux_x64` which is in the same as the other jobs in build?
- Pulling this sort of things directly from jenkins jobs makes me a bit nervous - I'm not sure it's ideal to have jenkins jobs be the master location of anything so perhaps we could have some other way of storing them. I've always preferred the `userContent` directory on jenkins 
- Given that this is a large download I'd personally prefer if we could incorporate these into the playbooks for installation onto the machine. The exact versions would be tracked and checksums verified in the same way that we do for GCC at the moment.
How often are the devkits likely to change? Is the JDK version the best way to index them?

**Comment by andrew-m-leonard** [comment]
> A few thoughts:
> 
>     * Is there a reason why the job name is e.g. `x64Linux` and not `linux_x64` which is in the same as the other jobs in build?
> 
>     * Pulling this sort of things directly from jenkins jobs makes me a bit nervous - I'm not sure it's ideal to have jenkins jobs be the master location of anything so perhaps we could have some other way of storing them. I've always preferred the `userContent` directory on jenkins
> 
>     * Given that this is a large download I'd personally prefer if we could incorporate these into the playbooks for installation onto the machine. The exact versions would be tracked and checksums verified in the same way that we do for GCC at the moment.
>       How often are the devkits likely to change? Is the JDK version the best way to index them?

I chose the arch to match the build configuration :-) but we can choose which ever..

So devkit make dependencies are defined in the [make/devkit/Tools.gmk](https://github.com/openjdk/jdk21u/blob/36b5ac46ed4cb450a17fc993d2b734c0a8e32bc0/make/devkit/Tools.gmk#L98)
which in theory could change between versions, or at any time! So this point is for debate, I suspect the reality is the jdk21u gcc11.3.0 won't change, and is currently built the same for jdk22 and jdk23, the latter have added gcc13.2.0.
The really important thing is for 3rd party "reproducible" builds a 3rd party can build the identical devkit.

So I think we could reasonably build by "gcc version" as opposed to jdk version, but on making that decision say for example building jdk23 with gcc11.3.0 i'd want to ensure the jdk32 Tools.gmk is identical to the jdk21u one...
We also need to come up with a precise "versioning" method, ie.some metadata within the tarball that states something like:
```
gcc version output = gcc 11.3.0 - Centos 7.9
devkit openjdk source = https://github.com/openjdk/jdk21u/commit/36b5ac46ed4cb450a17fc993d2b734c0a8e32bc0
devkit_identifier = linux_x64-gcc11.3.0-04-Mar-2024-001
```
Then in the ci-jenkins-pipeline configuration the devkit is specified by the devkit_identifier "gcc11.3.0-linux_x64-04-Mar-2024-001", and then verified with the /usr/local/devkit/gcc11.3.0 content to ensure validity?
The devkit version metadata is then stored in the SBOM...

I'd be happy to putting the devkit's in userContent, although if we are going to use ansible playbooks to set them up, then they just need to be where ever they need to get it from?

@sxa let me know your thoughts?, thanks

**Comment by sxa** [comment]
> I chose the arch to match the build configuration :-) but we can choose which ever..

I think it's better if the jenkins job names are consistent so I'd go with that if possible. The other thing we could do is have the job as a multiconfiguration or pipeline job that builds all the platforms together, which would also keep the artifacts all in one job (so similar to the centos7_docker_image_creator or the old AdoptOpenJDK docker multiarch jobs.

> So devkit make dependencies are defined in the [make/devkit/Tools.gmk](https://github.com/openjdk/jdk21u/blob/36b5ac46ed4cb450a17fc993d2b734c0a8e32bc0/make/devkit/Tools.gmk#L98) which in theory could change between versions, or at any time! So this point is for debate, I suspect the reality is the jdk21u gcc11.3.0 won't change, and is currently built the same for jdk22 and jdk23, the latter have added gcc13.2.0.

> we could reasonably build by "gcc version" as opposed to jdk version, but on making that decision say for example building jdk23 with gcc11.3.0 i'd want to ensure the jdk32 Tools.gmk is identical to the jdk21u one... We also need to come up with a precise "versioning" method, ie.some metadata within the tarball that states something like:

Are we able to add in an identifier to the version string? (There's a separate question about "should we"?). At the moment we have `Adoptium/SXA` in many of the GCC builds - could we have `Adoptium-b01/jdk21/sha` or `Adoptium-devkit21-sha` to indicate which we built from, and have b01 be a parameter we could increment if we needed to rebuild with the same gcc level but some other changes? That could go into the resultant tarballs (I'd also suggest using `xz` to compress instead of `gzip` since that will save a pile of space).

> The really important thing is for 3rd party "reproducible" builds a 3rd party can build the identical devkit.

Agreed.

> ```
> gcc version output = gcc 11.3.0 - Centos 7.9
> devkit openjdk source = https://github.com/openjdk/jdk21u/commit/36b5ac46ed4cb450a17fc993d2b734c0a8e32bc0
> devkit_identifier = linux_x64-gcc11.3.0-04-Mar-2024-001
> ```

Is that particular devkit identifier somethign under our control or is that a standard format that pops out somewhere?

> I'd be happy to putting the devkit's in userContent, although if we are going to use ansible playbooks to set them up, then they just need to be where ever they need to get it from?

I think I'd prefer that to jenkins jobs, just because it's a bit nicer if someone else wants to make use of our stuff to decouple it a bit from jenkins. Another option would be to have a devkit-binaries release repository and store them in there similar to what we do with temurin itself.

**Comment by andrew-m-leonard** [comment]
>The other thing we could do is have the job as a multiconfiguration or pipeline job that builds all the platforms together, which would also keep the artifacts all in one job (so similar to the centos7_docker_image_creator or the old AdoptOpenJDK docker multiarch jobs.

yeah, I was toying with that, if the jobs typically always build green then doing that would probably be best

> Are we able to add in an identifier to the version string? (There's a separate question about "should we"?). At the moment we have `Adoptium/SXA` in many of the GCC builds - could we have `Adoptium-b01/jdk21/sha` or `Adoptium-devkit21-sha` to indicate which we built from, and have b01 be a parameter we could increment if we needed to rebuild with the same gcc level but some other changes? That could go into the resultant tarballs (I'd also suggest using `xz` to compress instead of `gzip` since that will save a pile of space).

So no, as this makes the "devkit" non-reproducible, as this version info gets embedded in compiled obj/debuginfo, the devkit gcc -version is standard:
```
gcc (GCC) 11.3.0
```

> > ```
> > gcc version output = gcc 11.3.0 - Centos 7.9
> > devkit openjdk source = https://github.com/openjdk/jdk21u/commit/36b5ac46ed4cb450a17fc993d2b734c0a8e32bc0
> > devkit_identifier = linux_x64-gcc11.3.0-04-Mar-2024-001
> > ```
> 
> Is that particular devkit identifier somethign under our control or is that a standard format that pops out somewhere?

I've just made the above up! just as an indication of something we'd need, in essence to provide tracability and guarantee of what we're building with. So the devkit_identifier could be something relating to the "devkit build" it is from?

> I think I'd prefer that to jenkins jobs, just because it's a bit nicer if someone else wants to make use of our stuff to decouple it a bit from jenkins. Another option would be to have a devkit-binaries release repository and store them in there similar to what we do with temurin itself.

"devkit-gccNN-binaries" could be a good option. We could publish them as tags like "gcc-11.3.0-\<date\>", which would also be the devkit_identifier and the ci-jenkins pipeline config would reference by that tag...?

**Comment by andrew-m-leonard** [comment]
@sxa For the initial cycle I think i'm going to go with downloading the devkit (from devkit-gccNN-binaries or whereever..). In a 2nd stage we can implement ansible caching in /usr/local/bin, and if not found there it would do the 1st stage download...

**Comment by sxa** [comment]
> So no, as this makes the "devkit" non-reproducible, as this version info gets embedded in compiled obj/debuginfo, the devkit gcc -version is standard:

Unfortunate - it would be nice to be able to include something to indicate it was built from a devkit. So we can't have an identifiable build number in there.

> "devkit-gccNN-binaries" could be a good option. We could publish them as tags like "gcc-11.3.0-<date>", which would also be the devkit_identifier and the ci-jenkins pipeline config would reference by that tag...?

I'd just go with devkit-binaries (I can't see a good reason to create a new repo for each GCC version) and have release versions/tags for each devkit we build (including potentially the `-b01` etc. number so we can have multiple builds if required (e.g. the devkit makefiles were updated but still used the same gcc) that show up in the release name and `tar.xz` filename but not in the binaries.

Does that sound reasonable? Any gotchas you can think of?

**Comment by andrew-m-leonard** [comment]
> > So no, as this makes the "devkit" non-reproducible, as this version info gets embedded in compiled obj/debuginfo, the devkit gcc -version is standard:
> 
> Unfortunate - it would be nice to be able to include something to indicate it was built from a devkit. So we can't have an identifiable build number in there.
> 
> > "devkit-gccNN-binaries" could be a good option. We could publish them as tags like "gcc-11.3.0-", which would also be the devkit_identifier and the ci-jenkins pipeline config would reference by that tag...?
> 
> I'd just go with devkit-binaries (I can't see a good reason to create a new repo for each GCC version) and have release versions/tags for each devkit we build (including potentially the `-b01` etc. number so we can have multiple builds if required (e.g. the devkit makefiles were updated but still used the same gcc) that show up in the release name and `tar.xz` filename but not in the binaries.
> 
> Does that sound reasonable? Any gotchas you can think of?

I think that sounds reasonable, i'll work through this and see how it looks, thanks

**Comment by andrew-m-leonard** [comment]
Added a new "publish_devkit_tool" : https://github.com/adoptium/devkit-binaries/releases/tag/gcc-11.3.0-Centos7.6.1810-b01

File: pipelines/build/devkit/build_devkit.groovy [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ *   ppc64le
+ *   s390x
+ */
+
+def build_devkit() {
+    stage('Build DevKit') {
+        def openjdkRepo = "https://github.com/openjdk/${params.VERSION}.git"
+ 
+        // Clone upstream openjdk repo
+        sh(script:"git clone --depth 1 ${openjdkRepo} ${params.VERSION}")
```
[the remark this answers is not in the document]
Comment (andrew-m-leonard) [reply pipelines/build/devkit/build_devkit.groovy]: yep done, eg:
```
./make_devkit.sh jdk21u aarch64 Centos 7.6.1810
```

**Comment by andrew-m-leonard** [comment]
Added update to Tools.gmk to perform GPG verification on all the downloads.
MPFR cannot be gpg verified as Centos7 gpg 2.0.22 does not support the latest key format of the MPFR public gpg key. So a sha256 sum check is performed instead.

**Comment by andrew-m-leonard** [comment]
run tests

**Comment by andrew-m-leonard** [comment]
run tests

**Review by karianna — APPROVED** [review]

**Comment by sxa** [comment]
Looking at [the previous sha256.txt file](https://ci.adoptium.net/job/build-scripts/job/utils/job/devkit/job/gcc-11.3.0/job/devkit-gcc-11.3.0-linux-x64/lastSuccessfulBuild/artifact/workspace/devkit-gcc-11.3.0-Centos7.9.2009-x86_64-linux-gnu-b01.tar.xz.sha256.txt/*view*/) the filename in there has a `workspace/` prefix which makes it more difficult to use `sha256sum -c`. Can we remove the `workspace` prefix when it's generated?

This is a similar issue to the one for the windows 32-bit JDK8 builds from https://github.com/adoptium/temurin-build/issues/3662

**Review by sxa — APPROVED** [review]
I feel the build dir changes might have been better in a separate PR for ease of review.
I'm approving based on the fact I've run some of these jobs as part of the s390x testing and they appear to work :-) Some extra changes will be required for RHEL support but they will need to wait a bit pending some decisions on which environments to use Ref: https://github.com/adoptium/temurin-build/issues/3700

One question: Shat were the reasons for changing the binutils patch to not use `doc` explcitly?

**Comment by andrew-m-leonard** [comment]
> I feel the build dir changes might have been better in a separate PR for ease of review. I'm approving based on the fact I've run some of these jobs as part of the s390x testing and they appear to work :-) Some extra changes will be required for RHEL support but they will need to wait a bit pending some decisions on which environments to use Ref: [adoptium/temurin-build#3700](https://github.com/adoptium/temurin-build/issues/3700)
> 
> One question: Shat were the reasons for changing the binutils patch to not use `doc` explcitly?
> 
> EDIT: I'm really thinking it would be preferable to have a top level directory inside the devkit tarballs

Thank you @sxa 
binutils patch is a known binutils bug for Centos7 : https://sourceware.org/bugzilla/show_bug.cgi?id=29476

I think the devkit tarball structure I wanted to keep simple, without having to handle arch folder renaming/..

**Pull request merged**